### PR TITLE
Refactor to allow for sharing asterisk parsing w/ more parsers

### PR DESCRIPTION
**Pull request by CodeSmell**
in prep for ASN wanted to move some capability to interface

### CODE REVIEW

File: src/main/java/com/walmartlabs/x12/dex/dx894/DefaultDex894Parser.java [hunk 1]
```diff
@@ -611,13 +610,19 @@ protected String retreiveElementFromSegment(List<String> elements, int listIndex
...
+//    protected List<String> splitSourceDataIntoSegments(String sourceData) {
+//        return Arrays.asList(sourceData.split("\\r?\\n"));
+//    }
+//
+//    /**
+//     * moved to default method on X12Parser
+//     */
+//    protected List<String> splitSegmentIntoDataElements(String segment) {
+//        return Arrays.asList(segment.split("\\*"));
+//    }
```
Comment (patelch) [inline src/main/java/com/walmartlabs/x12/dex/dx894/DefaultDex894Parser.java]: Did you mean to leave this code here commented out?

File: src/main/java/com/walmartlabs/x12/dex/dx894/DefaultDex894Parser.java [hunk 2]
```diff
@@ -627,12 +632,12 @@ protected String retreiveElementFromSegment(List<String> elements, int listIndex
...
     protected String segmentIdentifier(int segmentIdx, List<String> dexSegments) {
         String segment = dexSegments.get(segmentIdx);
-        return this.segmentIdentifier(segment);
+        return this.extractSegmentIdentifier(segment);
     }
 
-    protected String segmentIdentifier(String segment) {
-        return segment.substring(0, segment.indexOf("*"));
-    }
+//    protected String extractSegmentIdentifier(String segment) {
```
Comment (patelch) [inline src/main/java/com/walmartlabs/x12/dex/dx894/DefaultDex894Parser.java]: Same here

File: src/main/java/com/walmartlabs/x12/X12Parser.java [hunk 3]
```diff
@@ -17,14 +17,37 @@
...
+     */
+    default List<String> splitSegmentIntoDataElements(String segment) {
+        return Arrays.asList(segment.split("\\*"));
+    }
+
+    /**
+     * extracts the first data element in a segment which is the segment identifier
+     */
+    default String extractSegmentIdentifier(String segment) {
+        return segment.substring(0, segment.indexOf("*"));
```
Comment (jaympatel) [inline src/main/java/com/walmartlabs/x12/X12Parser.java]: `segment.indexOf("*")` could return `-1` if * is not part of the string. Should we put check?

File: src/main/java/com/walmartlabs/x12/X12Parser.java [hunk 4]
```diff
@@ -17,14 +17,37 @@
...
+     */
+    default List<String> splitSourceDataIntoSegments(String sourceData) {
+        return Arrays.asList(sourceData.split("\\r?\\n"));
+    }
+
+    /**
+     * parses the segment into a list of data elements each date element is separated by an asterisk (*)
+     */
+    default List<String> splitSegmentIntoDataElements(String segment) {
+        return Arrays.asList(segment.split("\\*"));
```
Comment (jaympatel) [inline src/main/java/com/walmartlabs/x12/X12Parser.java]: Should we put check that `segment` has `*` or not?

File: src/test/java/com/walmartlabs/x12/X12ParserTest.java [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        byte[] dexBytes = Files.readAllBytes(Paths.get("src/test/resources/dex/894/dex.sample.1.txt"));
+        List<String> segmentsList = defaultParser.splitSourceDataIntoSegments(new String(dexBytes));
+        assertNotNull(segmentsList);
+        assertEquals(22, segmentsList.size());
+        assertEquals("DXS*9251230013*DX*004010UCS*1*9254850000", segmentsList.get(0));
+        assertEquals("DXE*1*2", segmentsList.get(21));
+    }
+
+    @Test
+    public void test_splitSourceDataIntoSegments_Empty() throws IOException {
```
Comment (patelch) [inline src/test/java/com/walmartlabs/x12/X12ParserTest.java]: This test won't throw an IOException, right?

File: src/test/java/com/walmartlabs/x12/X12ParserTest.java [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+        List<String> segmentsList = defaultParser.splitSourceDataIntoSegments("");
+        assertNotNull(segmentsList);
+        assertEquals(0, segmentsList.size());
+    }
+
+    @Test
+    public void test_splitSourceDataIntoSegments_Null() throws IOException {
```
Comment (patelch) [inline src/test/java/com/walmartlabs/x12/X12ParserTest.java]: Same here

**Pull request merged**